### PR TITLE
Add new rule to add annotations from router configuration to symfony controllers

### PR DESCRIPTION
**Pull request by malteschlueter**
See https://github.com/rectorphp/rector-symfony/issues/168

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Btw, I've triggered the CI for feedback. Could you check before review?

File: src/Command/RectorRoutesExportCommand.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+use Symfony\Bundle\FrameworkBundle\Console\Helper\DescriptorHelper;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * This command must be added to the application to export the routes to work with the "AddRouteAnnotationRector" rule
```
Comment (malteschlueter) [inline src/Command/RectorRoutesExportCommand.php:15]: There exists some files with information about routes inside the `var/cache` directory but the information is not direct accessible and the controllers are not resolved. There still stand `FooBundle:Bar:baz`.

So I tried to use the symfony command `debug:router --format=json` but there are also the controllers not resolved. But if you use this command inside the application it resolves the controllers correct.

TBD

File: src/Rector/ClassMethod/AddRouteAnnotationRector.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+        if (count($route->getMethods()) > 0) {
+            $items['methods'] = new CurlyListNode($route->getMethods());
+        }
+
+        if ($route->getCondition() !== '') {
+            $items['condition'] = $route->getCondition();
+        }
+
+        $phpDocInfo->addTagValueNode($this->symfonyRouteTagValueNodeFactory->createFromItems($items));
```
Comment (malteschlueter) [inline src/Rector/ClassMethod/AddRouteAnnotationRector.php:89]: So here is still my problem of the missing quotes. I've managed to add a first test to show you the problem

**Comment by malteschlueter** [comment]
> Btw, I've triggered the CI for feedback. Could you check before review?

The PHPUnit fails now only because of the new test I've added where you see the quote issue.

The PHPStan erros I will fix later

File: src/Rector/ClassMethod/AddRouteAnnotationRector.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        $phpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($node);
+        $routeAnnotation = $phpDocInfo->getByAnnotationClass(\Symfony\Component\Routing\Annotation\Route::class);
+
+        if ($routeAnnotation !== null) {
+            return null;
+        }
+
+        $items = [
+            'path' => $route->getPath(),
+            'name' => $route->getName(),
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply src/Rector/ClassMethod/AddRouteAnnotationRector.php:64]: I'm checking the Rector annotation factory implementation, and this seems the way to go.

The reason is that we always parse existing code as it is. Annotation cannot know if the value is string, int, bool or class constant refernce, as everything is string for the annotatoin parser.

```php
/**
 * @Route(key="value", anotherKey=1, yetAnotherKey=SomeClass::CONSTANT_REFERENCE, also="SomeKey::Value")
 */
```

We could add some assuming, but that might create magic conflicts, mainly for string values with :: in it.
So better create use the quotes for now.

[the remark this answers is not in the document]
Comment (malteschlueter) [reply src/Rector/ClassMethod/AddRouteAnnotationRector.php:64]: Alright, I will add the quotes by hand

File: src/ValueObject/Route.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Symfony\ValueObject;
+
+class Route
```
[the remark this answers is not in the document]
Comment (malteschlueter) [reply src/ValueObject/Route.php:7]: No problem. I will change it

File: src/Command/RectorRoutesExportCommand.php (hunk 1 above)
Comment (malteschlueter) [reply src/Command/RectorRoutesExportCommand.php:15]: Could work but I'm failing :/ Maybe that can do someone smarter than me

Comment (TomasVotruba) [reply src/Command/RectorRoutesExportCommand.php:14]: I see. Thanks for trying all the options. We can merge this PR, if CI is passing. The code does not have to be complete.

Could you share a minimal reproducible repository with your setup?
I see what can we do.

Comment (malteschlueter) [reply src/Command/RectorRoutesExportCommand.php:14]: I've created a minimal Symfony 3.4 application which should help to debug.
https://github.com/malteschlueter/symfony-reproducers/tree/feature/rector-add-route-annotation

That is what I've changed in Rector to require the container
```diff
diff --git a/src/ValueObjectFactory/RouteMapFactory.php b/src/ValueObjectFactory/RouteMapFactory.php
index 8adee63..5681cbc 100644
--- a/src/ValueObjectFactory/RouteMapFactory.php
+++ b/src/ValueObjectFactory/RouteMapFactory.php
@@ -17,6 +17,8 @@ final class RouteMapFactory
 
     public function createFromFileContent(string $configFilePath): RouteMap
     {
+        require $configFilePath;
+
         $fileContents = $this->smartFileSystem->readFile($configFilePath);
         $json = json_decode($fileContents, true);
 
diff --git a/tests/Rector/ClassMethod/AddRouteAnnotationRector/config/configured_rule.php b/tests/Rector/ClassMethod/AddRouteAnnotationRector/config/configured_rule.php
index 8fbc08d..e98921c 100644
--- a/tests/Rector/ClassMethod/AddRouteAnnotationRector/config/configured_rule.php
+++ b/tests/Rector/ClassMethod/AddRouteAnnotationRector/config/configured_rule.php
@@ -8,5 +8,5 @@ use Rector\Symfony\Rector\ClassMethod\AddRouteAnnotationRector;
 return static function (RectorConfig $rectorConfig): void {
     $rectorConfig->import(__DIR__ . '/../../../../../config/config.php');
     $rectorConfig->rule(AddRouteAnnotationRector::class);
-    $rectorConfig->symfonyRoutesJson(__DIR__ . '/route.json');
+    $rectorConfig->symfonyRoutesJson('/Users/mschlueter/Projects/Privat/symfony-reproducers/var/cache/dev/AppDevDebugProjectContainer.php');
 };
```

Comment (malteschlueter) [reply src/Command/RectorRoutesExportCommand.php:14]: PHPStan says it doesn't know `Symfony\Component\Routing\RouterInterface::getRouteCollection()` but the component is installed. See https://github.com/rectorphp/rector-symfony/runs/6442259714?check_suite_focus=true#step:4:124
How does I solve it?

Unfortunately I've no idea what this job is doing and why it's failing :D Any idea? https://github.com/rectorphp/rector-symfony/runs/6442259666?check_suite_focus=true

Comment (TomasVotruba) [reply src/Command/RectorRoutesExportCommand.php:14]: > Unfortunately I've no idea what this job is doing and why it's failing :D Any idea?

Here it's Github UX fail. Check the 2 jobs above, this one is changing code:

```bash
vendor/bin/rector process --ansi
```

Best way to fix is running it locally and commit :)

Comment (malteschlueter) [reply src/Command/RectorRoutesExportCommand.php:14]: And thank you very much for your help @TomasVotruba :)
Hope we can finish this feature and it works well for the most people (everyone will take some time I think :D )
I think it can help many people who works on legacy applications and wants to update to annotation/attributes

Comment (TomasVotruba) [reply src/Command/RectorRoutesExportCommand.php:14]: > PHPStan says it doesn't know Symfony\Component\Routing\RouterInterface::getRouteCollection() but the component is installed. See [rectorphp/rector-symfony/runs/6442259714?check_suite_focus=true#step:4:124](https://github.com/rectorphp/rector-symfony/runs/6442259714?check_suite_focus=true#step:4:124)
How does I solve it?


You can skip this file in `phpstan.neon`, as it requires external context

Comment (TomasVotruba) [reply src/Command/RectorRoutesExportCommand.php:14]: > And thank you very much for your help @TomasVotruba :)

You're welcome :+1:  

I have no doubts, we'll finish it this week to test it out on edge-cases. You've done great works so far, thank you :bow:

Comment (TomasVotruba) [reply src/Command/RectorRoutesExportCommand.php:14]: I'm in the https://github.com/malteschlueter/symfony-reproducers/tree/feature/rector-add-route-annotation in `rector-add-route-annotation` branch. But when I run:

```bash
vendor/bin/rector
```

it's missing the package. What exact steps should I do to make Rector fail?

Comment (TomasVotruba) [reply src/Command/RectorRoutesExportCommand.php:14]: Btw, after Rector + PHPStan fix in CI, we can merge it to progress with testing on `rector/rector-dev-main` :wink:

**Comment by TomasVotruba** [comment]
Could you rebase on `dev-main`? I've cherry picked 2 changes.

**Comment by TomasVotruba** [comment]
Let's merge this :+1:
